pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -106,6 +121,70 @@ func haveOverlap(a1, a2 []string) bool {
 	return false
 }
 
+func (pl *VolumeRestrictions) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
+	if pl.enableReadWriteOncePod {
+		return pl.isReadWriteOncePodAccessModeConflict(pod)
+	}
+	return framework.NewStatus(framework.Success)
+}
+
+// isReadWriteOncePodAccessModeConflict checks if a pod uses a PVC with the ReadWriteOncePod access mode.
+// This access mode restricts volume access to a single pod on a single node. Since only a single pod can
+// use a ReadWriteOncePod PVC, mark any other pods attempting to use this PVC as UnschedulableAndUnresolvable
+// (alpha).
+// For the ReadWriteOncePod beta, pods with conflicts will be marked Unschedulable.
+func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(pod *v1.Pod) *framework.Status {
+	nodeInfos, err := pl.nodeInfoLister.NodeInfos().List()
+	if err != nil {
+		return framework.NewStatus(framework.Error, "error while getting node info")
+	}
+
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvc, err := pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(volume.PersistentVolumeClaim.ClaimName)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonPVCNotFound)
+			}
+			return framework.NewStatus(framework.Error, "unexpected error while getting PersistentVolumeClaim")
+		}
+
+		// If ReadWriteOncePod is used, it can be the only one on the PVC since it overrides any
+		// additional access modes. This is also because we do not know what access mode the pod
+		// intends to use if there are multiple.
+		if !(v1helper.ContainsAccessMode(pvc.Spec.AccessModes, v1.ReadWriteOncePod) && len(pvc.Spec.AccessModes) == 1) {
+			continue
+		}
+
+		key := pod.Namespace + "/" + volume.PersistentVolumeClaim.ClaimName
+		status := framework.NewStatus(framework.Success)
+
+		processNode := func(i int) {
+			nodeInfo := nodeInfos[i]
+			refCount := nodeInfo.PVCRefCounts[key]
+			// Enforce ReadWriteOncePod access mode. This is also enforced during volume mount in kubelet.
+			if refCount > 0 {
+				// Preemption is not supported for the ReadWriteOncePod alpha; it is targeted for beta.
+				status = framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonReadWriteOncePodConflict)
```

Comment:
are we assuming that one and only one node should execute this path? if so please add a comment.

Even if multiple threads try to set status/found, we are still ok if we don't use a mutex because we are fine with any value, is that correct (in C/C++ this is safe to do with primitive values, but I am not sure about golang)?

Reply:
I think it's safer to use the atomic library. But better use a boolean as suggested above.